### PR TITLE
armv8m: Add MCX N94x target, fix armv8m build

### PR DESCRIPTION
**Pull request by agkaminski**
JIRA: RTOS-881

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (refactoring, style fixes, git/CI config, submodule management, no code logic changes)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (mcxn94x).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [x] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
https://github.com/phoenix-rtos/phoenix-rtos-project/pull/1120
https://github.com/phoenix-rtos/plo/pull/337
https://github.com/phoenix-rtos/phoenix-rtos-utils/pull/222
https://github.com/phoenix-rtos/phoenix-rtos-hostutils/pull/55
https://github.com/phoenix-rtos/phoenix-rtos-filesystems/pull/132
https://github.com/phoenix-rtos/phoenix-rtos-devices/pull/508
https://github.com/phoenix-rtos/phoenix-rtos-build/pull/195
https://github.com/phoenix-rtos/libphoenix/pull/371
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: hal/armv8m/mcx/n94x/mcxn94x.c [hunk 1]
```diff
@@ -0,0 +1,695 @@
...
+	else {
+		*(reg + (syscon_presetctrlcrl0 - syscon_presetctrl0)) = 1 << (dev & 0x1f);
+	}
+	hal_cpuDataMemoryBarrier();
+
+	return 0;
+}
+
+
+int hal_platformctl(void *ptr)
```
Comment (lukileczo) [inline hal/armv8m/mcx/n94x/mcxn94x.c:569]: `pltctlSp` spinlock is not used here

File: hal/armv8m/mcx/n94x/mcxn94x.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+extern u64 _mcxn94x_sysconGray2Bin(u64 gray);
+
+
+extern int _mcxn94x_sysconSetDevClk(int dev, unsigned int sel, unsigned int div, int enable);
+
+
+extern int _mcxn94x_sysconDevReset(int dev, int state);
+
+
+extern int hal_platformctl(void *);
```
Comment (lukileczo) [inline hal/armv8m/mcx/n94x/mcxn94x.h:45]: Parameter name missing

File: hal/armv8m/mcx/interrupts.c [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+#include "hal/list.h"
+#include "hal/armv8m/armv8m.h"
+#include "hal/armv8m/mcx/n94x/mcxn94x.h"
+
+#include "proc/userintr.h"
+
+#include "config.h"
+
+/* Value based on other target architectures */
+#define SIZE_HANDLERS 4
```
Comment (lukileczo) [inline hal/armv8m/mcx/interrupts.c:28]: Unused define, I guess it's a leftover that wasn't removed (present in most irq drivers). General comment: it'd be nice to share the NVIC code across CPUs using it, but that's not in scope of this PR.

File: hal/armv8m/Makefile [hunk 4]
```diff
@@ -10,4 +10,9 @@ ifneq (, $(findstring nrf, $(TARGET_SUBFAMILY)))
     CFLAGS += -Ihal/armv8m
 endif
 
+ifneq (, $(findstring mcx, $(TARGET_SUBFAMILY)))
```
Comment (lukileczo) [inline hal/armv8m/Makefile:13]:
```suggestion
else ifneq (, $(findstring mcx, $(TARGET_SUBFAMILY)))
```

File: hal/armv8m/cpu.c [hunk 5]
```diff
@@ -182,8 +182,10 @@ char *hal_cpuInfo(char *info)
 	int i;
 	unsigned int cpuinfo;
 
-#ifdef CPU_NRF91
+#if defined(__CPU_NRF9160)
 	cpuinfo = _nrf91_cpuid();
+#elif defined(__CPU_MCXN94X)
+	cpuinfo = _mcxn94x_cpuid();
```
Comment (lukileczo) [inline hal/armv8m/cpu.c:188]: How about renaming these to e.g. `platform_cpuid` to remove the `#ifdefs`?

Comment (agkaminski) [reply hal/armv8m/cpu.c:188]: I'm not convinced, as this function is arch specific (it returns ARM-m related core revision). I'd rather unify SCB handling instead, perhaps later

**Review by lukileczo — DISMISSED** [review]
LGTM

File: hal/armv8m/mcx/_init.S [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+.globl _interrupts_dispatch
+.type _interrupts_dispatch, %function
+_interrupts_dispatch:
+	mov r0, #0
+	msr control, r0
+	isb
+
+	mov r0, sp
+	and r12, lr, #(1 << 2)
+	cmp r12, #0
```
Comment (kemonats) [inline hal/armv8m/mcx/_init.S:191]: Simpler
```suggestion
	ands r12, lr, #(1 << 2)
```
If the value of r12 is not used then
```suggestion
	tst  lr, #(1 << 2)
```

File: hal/armv8m/mcx/_init.S [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+_intd0:
+	bl interrupts_dispatch
+	ldr r1, [sp]
+	mov sp, r1
+	isb
+_intd1:
+	ldr lr, [sp, #(11 * 4)]
+
+	and r12, lr, #(1 << 2)
+	cmp r12, #0
```
Comment (kemonats) [inline hal/armv8m/mcx/_init.S:221]: as above

File: hal/armv8m/mcx/_init.S [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+	dmb
+	strexb r4, r3, [r1]
+	cmp r4, #0
+	bne spinlock
+	pop {r3-r4}
+hal_cpuReschedule0:
+	cpsie if
+	isb
+	dmb
+	bx lr
```
Comment (kemonats) [inline hal/armv8m/mcx/_init.S:284]: What do you think about such a solution?
```suggestion
/* int hal_cpuReschedule(struct _spinlock_t *spinlock, spinlock_ctx_t *scp); */
/*  scp variable ( r1 register ) is not used, you do not need to keep it */
hal_cpuReschedule:
	push {r4, lr}
	mov r0, r4
	bl _hal_invokePendSV
	cmp r4, #NULL
	beq .Lhal_cpuReschedule0
	add r4, r4, #12

.Lspinlock:
	ldrexb r2, [r4]
	add r2, r2, #1
	dmb
	strexb r3, r2, [r4]
	cmp r3, #0
	bne .Lspinlock
.Lhal_cpuReschedule0:
	cpsie if
	isb
	dmb
	movs r0, #0 /* default return value */
	pop {r4, lr}
```
In addition, you can use `cbz` and `adds` instructions ( smaller resulting code ).

File: hal/armv8m/spinlock.c [hunk 9]
```diff
@@ -26,21 +26,21 @@ void hal_spinlockSet(spinlock_t *spinlock, spinlock_ctx_t *sc)
...
-		strexb r2, r3, [%1]; \
-		cmp r2, #0; \
+		strexb r1, r2, [%1]; \
+		cmp r1, #0; \
 		bne 1b; \
 		dmb"
 	:
 	: "r" (sc), "r" (&spinlock->lock)
-	: "r2", "r3", "memory", "cc");
+	: "r1", "r2", "memory", "cc");
```
Comment (kemonats) [inline hal/armv8m/spinlock.c:43]: Why you change registers?
In your version, the compiler adds an instruction at the beginning of the function:
```
mov r3,r1
```
Same in `hal_spinlockClear()`

Comment (kemonats) [inline hal/armv8m/spinlock.c:43]: Why you change registers?
In your version, the compiler adds an instruction at the beginning of the function:

```
mov r3,r1
```
Same in `hal_spinlockClear()`

File: hal/armv8m/mcx/_init.S (hunk 6 above)
Comment (agkaminski) [reply hal/armv8m/mcx/_init.S]: Thanks, fixed

File: hal/armv8m/mcx/_init.S (hunk 8 above)
Comment (agkaminski) [reply hal/armv8m/mcx/_init.S]: I like your version, will update armv7m with it later, applied

File: hal/armv8m/spinlock.c (hunk 9 above)
Comment (agkaminski) [reply hal/armv8m/spinlock.c]: I assumed that armv7m version is fresher and used it to replace existing one, will update armv7m later, thanks

File: hal/armv8m/mcx/_init.S [hunk 10]
```diff
@@ -0,0 +1,355 @@
...
+	mrs r0, ipsr
+	mrs r3, psp
+	sub r1, sp, #48
+	stmdb sp!, {r1-r11, lr}
+
+	/* if we came from userspace, copy hw-saved regs to kstack
+	 * in case of signal handling
+	 */
+
+	beq _intd0 /* cmp r12, #0 */
```
Comment (lukileczo) [inline hal/armv8m/mcx/_init.S:204]:
```suggestion
	beq _intd0 /* tst lr, #(1 << 2) */
```

Comment (agkaminski) [reply hal/armv8m/mcx/_init.S]: Thx

**Review by lukileczo — APPROVED** [review]

**Pull request merged**